lib/Parser/Parser.cpp

```diff
@@ -32,7 +32,7 @@ Optional<PredicateDecl> Parser::parsePredicateDecl() {
     };
 
     if(identifier.type != Token::Type::identifier) {
-        emitSyntaxError("Expected predicate name in predicate definition.");
+        throw(SyntaxError("Expected predicate name in predicate definition."));
         return rewindAndReturn();
```

Comment:
I intended for (a subset?) of these parsing APIs to be "public." This code is built as a library called `AlliumParse`, which is dynamically linked to the Allium interpreter. However, the Allium language server will soon be the second user of the library, which will (eventually) need to perform incremental parsing using these APIs. I want to leave the door open for future users too; for example, static analyzers, linters, metaprogramming utilities, and any tool that needs a deep understanding of Allium programs.

Using exceptions for control flow out of the public interface is probably fine, but did you consider any alternatives? For example, is there a way to make the failure explicit in the return type (e.g. Swift's `Result` type)? What if we want to be able to issue multiple syntax errors someday?

For the record, I have 0 concerns about API/ABI stability given that Allium has 0 users, so we have a lot of latitude here.

Reply:
Looking back on it now, I'm not surprised that my current implementation isn't particularly future-proof. I initially avoided that approach of making the syntax errors part of the return type since it looked like it would require relatively major refactoring and I wasn't familiar with the existing code, but I'm comfortable enough with the parser code now that I can attempt it if that design makes sense to you. Alternatively, if you come up with a solution you like better I'm happy to do the legwork on the implementation of your preferred design.

Reply:
Yeah, after giving it some more thought, the solution I keep circling back to would work as follows:
* Define a type `Result<Success, Failure>` which is basically an alias of `TaggedUnion<Success, Failure>`
* Each parser method `parseX(...)` will return a `Result<X, std::vector<SyntaxError>>`
* Each parser method should recursively accumulate syntax errors. This should have acceptable performance since programs typically won't have enormous numbers of syntax errors at once (currently we diagnose at most one)

It should be straightforward to write a utility to combine error results (it's just concatenation!), but it might take some creativity to come up with a good abstraction to keep things concise in the parsers.

If it would be helpful, we could chat about this when you have a chance. I'm also always happy if you have other ideas.